File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.cpp
class RandomizedSet {
private:
    std::vector<int> _vec;
    std::unordered_map<int, int> _val_idx;
public:
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if (_val_idx.find(val) != _val_idx.end()){
            return false;
        }
        _vec.push_back(val);
        _val_idx[val] = _vec.size() - 1;
        return true;
    }
    
    bool remove(int val) {
        if (_val_idx.find(val) == _val_idx.end()) {
            return false;
        }
        int val_idx = _val_idx[val];
        int lst_val = _vec.back();
        _vec[val_idx] = lst_val;
        _val_idx[lst_val] = val_idx;
        _vec.pop_back();
        _val_idx.erase(val);
        return true;
    }
    
    int getRandom() {
        return _vec[std::rand() % _vec.size()];
    }
};

File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.py
class RandomizedSet:

    def __init__(self):
        self.store = []
        self.map = {}

    def insert(self, val: int) -> bool:
        if val in self.map:
            return False
        self.store.append(val)
        self.map[val] = len(self.store) - 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.map:
            return False
        idx = self.map[val]
        lst_val = self.store[-1]
        self.store[idx] = lst_val
        self.map[lst_val] = idx
        del self.map[val]
        self.store.pop()
        return True

    def getRandom(self) -> int:
        return random.choice(self.store)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()